CommandLine/libGMX/gmx.cpp

```diff
@@ -0,0 +1,382 @@
+  }
+}
+
+void PackRes(std::string &dir, std::string &name, int id, pugi::xml_node &node, google::protobuf::Message *m, int depth) {
+  const google::protobuf::Descriptor *desc = m->GetDescriptor();
+  const google::protobuf::Reflection *refl = m->GetReflection();
+  for (int i = 0; i < desc->field_count(); i++) {
+    const google::protobuf::FieldDescriptor *field = desc->field(i);
+    const google::protobuf::OneofDescriptor *oneof = field->containing_oneof();
+    if (oneof && refl->HasOneof(*m, oneof))
+      continue;
+    const google::protobuf::FieldOptions opts = field->options();
+
+    if (field->name() == "id") {
+      id += opts.GetExtension(buffers::id_start);
+      outputStream << "Setting " << field->name() << " (" << field->type_name() << ") as " << id << std::endl;
+      refl->SetInt32(m, field, id);
+    } else if (depth == 0 && field->name() == "name") {
+      outputStream << "Setting " << field->name() << " (" << field->type_name() << ") as " << name.c_str() << std::endl;
+      refl->SetString(m, field, name);
+    } else {
+      const std::string gmxName = opts.GetExtension(buffers::gmx);
+      const bool isFilePath = opts.GetExtension(buffers::file_path);
+      std::string xmlElement = gmxName;
+      bool isSplit = false;
+      bool isAttribute = false;
+      pugi::xml_node child = node;
+      pugi::xml_attribute attr;
+
+      // if deprecated we'll just try gmx name anyway and not whine when it fails
+      if (gmxName == "GMX_DEPRECATED")
+        xmlElement = gmxName;
+      else {
+        // use the name the protobuf field uses unless there a (gmx) attr
+        if (xmlElement.empty()) xmlElement = field->name();
+
+        // this is for <point>0,0</point> crap
+        const std::string splitMarker = "GMX_SPLIT/";
+        size_t splitPos = gmxName.find(splitMarker);
+        isSplit = splitPos != std::string::npos;
+
+        // if it's not a split then we deal with yoyo's useless nesting
+        if (!isSplit) {
+          std::vector<std::string> nodes = SplitString(xmlElement, '/');
+          for (auto n : nodes) {
+            child = child.child(n.c_str());
+            child.append_attribute("visited") = "true";
+            xmlElement = n;
+          }
+        }
+
+        // We want the data from the node "child" but if its empty what we seek is likely in the attributes
+        if (child.empty()) attr = node.attribute(xmlElement.c_str());
+        isAttribute = !attr.empty();
+
+        if (child.empty() && !isAttribute && !field->is_repeated()) {
+          // ename only exists if etype = 4. Also, etype and enumb don't exist in timeline events
+          pugi::xml_attribute a = node.attribute("eventtype");
+          if (xmlElement != "ename" && a.as_int() != 4 && node.path() != "/timeline/entry/event")
+            outputStream << "Error: no such element " << node.path() << "/" << xmlElement << std::endl;
+        } else {
+          if (field->is_repeated()) {
```

Comment:
This block seems to be logic to translate Pugi → Proto. Can you extract it into its own method?

Reply:
moving this shit requires a function that takes ~50 variables and makes things harder to read.

Reply:
That's fine. Maybe add a comment to each branch (if/else block) explaining what the branch does and why it was taken.

Reply:
It would also be possible to make a context struct and pass the context around. I'm not too worried about this, this function is pretty short, and it's also like the only thing to this file. Obviously it's going to need some refactoring to make it double as a JSON deserializer for GMS2 projects. So we can just wait until then to break this thing down farther.